cirq-google/cirq_google/ops/common_gate_families.py

```diff
@@ -0,0 +1,362 @@
+            f'allow_symbols={self.allow_symbols},'
+            f'atol={self.atol})'
+        )
+
+    def _value_equality_values_(self) -> Any:
+        return (
+            frozenset(self.gates_to_accept),
+            frozenset(self.gate_types_to_check),
+            self.allow_symbols,
+            self.atol,
+        )
+
+    def _approx_eq_or_symbol(self, lhs: Any, rhs: Any) -> bool:
+        lhs = lhs if isinstance(lhs, tuple) else (lhs,)
+        rhs = rhs if isinstance(rhs, tuple) else (rhs,)
+        assert len(lhs) == len(rhs)
+        for l, r in zip(lhs, rhs):
+            is_parameterized = cirq.is_parameterized(l) or cirq.is_parameterized(r)
+            if (is_parameterized and not self.allow_symbols) or (
+                not is_parameterized and not cirq.approx_eq(l, r, atol=self.atol)
+            ):
+                return False
+        return True
+
+    def _get_value_equality_values(self, g: POSSIBLE_FSIM_GATES) -> Any:
+        # TODO: Remove condition once https://github.com/quantumlib/Cirq/issues/4585 is fixed.
+        if type(g) == cirq.PhasedISwapPowGate:
+            return (g.phase_exponent, *g._iswap._value_equality_values_())  # type: ignore
+        return g._value_equality_values_()
+
+    def _get_value_equality_values_cls(self, g: POSSIBLE_FSIM_GATES) -> Any:
+        # TODO: Remove condition once https://github.com/quantumlib/Cirq/issues/4585 is fixed.
+        if type(g) == cirq.PhasedISwapPowGate:
+            return cirq.PhasedISwapPowGate
+        return g._value_equality_values_cls_()  # type: ignore
+
+    def _check_equal(self, g1: POSSIBLE_FSIM_GATES, g2: POSSIBLE_FSIM_GATES) -> bool:
+        if not self.allow_symbols:
+            return g1 == g2 and not (cirq.is_parameterized(g1) or cirq.is_parameterized(g2))
+        if self._get_value_equality_values_cls(g1) != self._get_value_equality_values_cls(g2):
+            return False
+        return self._approx_eq_or_symbol(
+            self._get_value_equality_values(g1), self._get_value_equality_values(g2)
+        )
+
+    def _predicate(self, gate: cirq.Gate) -> bool:
+        if not isinstance(gate, self.gate_types_to_check):
+            return False
+        gate = cast(POSSIBLE_FSIM_GATES, gate)
+        for g in self.gates_to_accept:
+            if isinstance(g, type):
+                cg = self.convert(gate, g)
+                if cg is not None:
+                    return True
+            else:
+                for target in type(gate).mro():
+                    if target in self.gate_types_to_check:
+                        cg = self.convert(g, target)
+                        if cg is None:
+                            continue
+                        if self._check_equal(gate, cg):
+                            return True
+                        break
+        return False
+
+    def convert(
+        self, gate: cirq.Gate, target_gate_type: Type[POSSIBLE_FSIM_GATES]
+    ) -> Optional[POSSIBLE_FSIM_GATES]:
```

Comment:
I wonder if we can fix the typing issue here, by using a type var, something like:
```python
from typing import TypeVar
...
T = TypeVar('T', bound=POSSIBLE_FSIM_GATES)
...
def convert(self, gate: cirq.Gate, target_gate_type: Type[T]) -> Optional[T]:
    ...
```
If this can work I think it would eliminate a bunch of casts when calling `convert`, but I'm not quite sure whether bounds on type vars work on Unions like this.

If the generic convert function can't be made typesafe, something we might want to do is make the _specific_ conversion functions public so they can be called in a type-safe way, e.g. if you know you want an `FSimGate` you could call `family.convert_to_fsim(gate)` instead of `cast(cirq.FSimGate, family.convert(gate, cirq.FSimGate))`.

Reply:
After getting stuck in the mypy hell for quite some time, I've made a change which seems to work and is typesafe at the callsite. Though there's an (unavoidable?) issue that if we call `convert(g, target)` where `target`'s type is `Union[Type[FSimGate], Type[ISwapPowGate], etc.]` (for example by iterating over a non-homogenous list of conversion targets), then mypy complaints. 

LMK if you have a better solution. I personally do prefer having a single `convert` method over having explicit methods for each target type.